// lebab/lib/transform/commonjs/matchNamedExport.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fMatches = require('f-matches');

var _isExports = require('./isExports');

var _isExports2 = _interopRequireDefault(_isExports);

var _isModuleExports = require('./isModuleExports');

var _isModuleExports2 = _interopRequireDefault(_isModuleExports);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Matches: exports.<id> = <value>
 * Matches: module.exports.<id> = <value>
 *
 * When match found, returns object with:
 *
 * - id
 * - value
 *
 * @param  {[type]} node [description]
 * @return {[type]}      [description]
 */
exports.default = (0, _fMatches.matches)({
  type: 'ExpressionStatement',
  expression: {
    type: 'AssignmentExpression',
    operator: '=',
    left: {
      type: 'MemberExpression',
      computed: false,
      object: function object(ast) {
        return (0, _isExports2.default)(ast) || (0, _isModuleExports2.default)(ast);
      },
      property: (0, _fMatches.extract)('id', {
        type: 'Identifier'
      })
    },
    right: (0, _fMatches.extractAny)('value')
  }
});
